### PR TITLE
Gpl177 create tube rack manifest

### LINKED ISSUE #2412: GPL-177 Generate Tube Rack Manifests
Add code to enable representation of tube racks in Sequencescape, and to generate tube rack manifests for Tree of Life.
Manifest generation to allow choice of rack size (tubes per rack), and number of racks.
Manifest to contain cells for entering one (or more) tube rack barcodes at the top of the spreadsheet, plus rows for tubes and their samples as in a normal tube manifest.
Tasks:
* Create Tube rack model and other required classes
* Create Config for manifest
* Create Tube rack Manifest code

### PR DESCRIPTION
**Pull request by KatyTaylor**
Am I merging into the correct branch here?

Just re-running all the tests now (RSpec) to check they all still work.

I added several tests that no-one's seen yet, and also did a bit of refactoring, so please review & all feedback welcome!

Thanks

closes #2412

### CODE REVIEW

File: lib/tasks/limber.rake [hunk 1]
```diff
@@ -122,6 +122,33 @@ namespace :limber do
...
+      )
+    end
+
+    unless Purpose.where(name: 'TR Stock 48').exists?
+      TubeRack::Purpose.create!(
+        name: 'TR Stock 48',
+        target_type: 'TubeRack',
+        stock_plate: true,
+        default_state: 'pending',
+        barcode_printer_type: BarcodePrinterType.find_by(name: '96 Well Plate'),
```
Comment (pjvv) [inline lib/tasks/limber.rake]: Are these supposed to be the same? Maybe a comment as to why they are?

File: spec/factories/purpose_factories.rb [hunk 2]
```diff
@@ -139,4 +139,14 @@
     sequence(:name) { |n| "Child tube purpose #{n}" }
     target_type { 'Tube' }
   end
+
+  factory :tube_rack_purpose_48, class: TubeRack::Purpose do
```
Comment (pjvv) [inline spec/factories/purpose_factories.rb]: The factory is named with '48' but has both sizes?

File: app/helpers/sdb/sample_manifests_helper.rb [hunk 3]
```diff
@@ -4,7 +4,8 @@ def count_labels
       '1dtube' => 'Tubes required',
       'plate' => 'Plates required',
       'library' => 'Tubes required',
-      'multiplexed_library' => 'Number of samples in library'
+      'multiplexed_library' => 'Number of samples in library',
+      'tube_rack' => 'Racks required'
```
[the remark this answers is not in the document]
Comment (KatyTaylor) [reply app/helpers/sdb/sample_manifests_helper.rb]: Yes, probably!

File: spec/factories/purpose_factories.rb (hunk 2 above)
Comment (KatyTaylor) [reply spec/factories/purpose_factories.rb]: Yeah I wasn't quite sure how to structure it, what would you do?

File: lib/tasks/limber.rake (hunk 1 above)
Comment (KatyTaylor) [reply lib/tasks/limber.rake]: What do you mean by these? The block above and this block? They have different names and sizes - 96 vs 48

File: spec/factories/purpose_factories.rb (hunk 2 above)
Comment (andrewsparkes) [reply spec/factories/purpose_factories.rb]: maybe the default 'tube_rack_purpose' factory is size 96, and we just override size and name if we want something else...

Comment (KatyTaylor) [reply spec/factories/purpose_factories.rb]: Sure, I'll do that.

**Comment by KatyTaylor** [comment]
Hi guys, I changed this pull request to go into our new tube rack feature branch, tube-racks-feature.

**Pull request merged**